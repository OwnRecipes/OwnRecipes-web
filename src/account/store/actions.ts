import jwtDecode, { JwtPayload } from 'jwt-decode';
import moment from 'moment';

import { refreshToken, request } from '../../common/CustomSuperagent';
import { serverURLs } from '../../common/config';
import { ACTION } from '../../common/store/ReduxHelper';
import LocalStorageHelper from '../../common/LocalStorageHelper';
import { AnyDispatch, toBasicAction } from '../../common/store/redux';
import { handleFormError } from '../../common/requestUtils';
import { AccountActionTypes, ACCOUNT_STORE, AccountDispatch, ACCOUNT_TOKEN_STORAGE_KEY, UserAccount, LoginDto, toUserAccount, AccountAction } from './types';

export function shouldForceReLogin(refresh: string): boolean {
  const decodedToken: JwtPayload = jwtDecode<JwtPayload>(refresh);
  return decodedToken.exp == null
      || (moment(new Date()).add(1, 'day') > moment.unix(decodedToken.exp));
}

export function shouldRefreshToken(token: string | undefined): boolean {
  const decodedToken: JwtPayload | undefined = token ? jwtDecode<JwtPayload>(token) : undefined;
  return decodedToken != null && decodedToken.exp != null
      && (moment(new Date()).add(2, 'minutes') > moment.unix(decodedToken.exp));
}

export const getToken = async (dispatch: AnyDispatch, username: string, pass: string, remember: boolean) => {
  dispatch({ ...toBasicAction(ACCOUNT_STORE, ACTION.UPDATE_START) });

  const url = serverURLs.auth_token;
  return request
    .post(url)
    .send({ username: username, password: pass })
    .then(res => {
      const data: LoginDto = res.body;
      dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.LOGIN), payload: toUserAccount(data, remember) });
      return null;
    })
    .catch(err => handleFormError(dispatch, err, ACCOUNT_STORE));
};

export const tryAutoLogin = () => (dispatch: AccountDispatch) => {
  const storageItem = LocalStorageHelper.getItem(ACCOUNT_TOKEN_STORAGE_KEY);
  if (storageItem == null) {
    return;
  }

  const user: UserAccount = JSON.parse(storageItem);
  if (!user.remember || !user.refresh || shouldForceReLogin(user.refresh)) {
    dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.FORGET_LOGIN) });
    return;
  }

  dispatch({ ...toBasicAction(ACCOUNT_STORE, ACTION.GET_START) });

  if (user.token) {
    if (shouldRefreshToken(user.token)) {
      refreshToken.instance(user.remember);
    } else {
      dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.LOGIN), payload: user });
    }
  } else {
    dispatch({ ...toBasicAction(ACCOUNT_STORE, ACTION.SOFT_RESET) });
  }
};

export const tryRefresh = () => (dispatch: AccountDispatch) => {
  const storageItem = LocalStorageHelper.getItem(ACCOUNT_TOKEN_STORAGE_KEY);
  if (storageItem == null) {
    return;
  }

  const user: UserAccount = JSON.parse(storageItem);
  if (!user.remember || !user.token) {
    return;
  }

  const decodedToken: JwtPayload | undefined = jwtDecode<JwtPayload>(user.token);
  if (decodedToken.exp && decodedToken.exp <= moment().unix()) {
    dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.FORGET_LOGIN) });
  } else if (shouldRefreshToken(user.token)) {
    refreshToken.instance(user.remember);
  } else {
    dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.LOGIN), payload: user });
  }
};

const authSideloadToken = (user: UserAccount): AccountAction => ({
  ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.SIDELOAD_TOKEN), payload: { ...user },
});

export const sideloadToken = (user: UserAccount) => (dispatch: AccountDispatch) => {
  dispatch(authSideloadToken(user));
};

export const forgetLogin = () => (dispatch: AccountDispatch) => {
  dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.FORGET_LOGIN) });
};

export const logUserOut = async (dispatch: AccountDispatch, oldToken: string | undefined) => {
  const url = serverURLs.revoke_token;
  return request
    .post(url)
    .send({ refresh: oldToken })
    .then(() => {
      dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.LOGOUT) });
      return null;
    })
    .catch(() => {
      dispatch({ ...toBasicAction(ACCOUNT_STORE, AccountActionTypes.LOGOUT) });
      return null;
    });
};
