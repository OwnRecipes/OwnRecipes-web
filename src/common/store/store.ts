/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line camelcase
import { applyMiddleware, compose, legacy_createStore } from 'redux';
import thunkMiddleware from 'redux-thunk';

import reducer from '../../app/Store';
import { isDemoMode } from '../utility';

let data;
if (isDemoMode()) {
  // eslint-disable-next-line global-require
  data = require('../../demo/store/data.json');
}

const composeEnhancers = (
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    && ((
      process.env.NODE_ENV === 'production'
        && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          features: {
            export: true,
          },
        })
    ) || (
      (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
        trace: true,
        traceLimit: 25,
      })
    ))
  ) || compose;

const store = legacy_createStore(
  reducer,
  data,
  composeEnhancers(
    applyMiddleware(thunkMiddleware)
  )
);

export default store;
