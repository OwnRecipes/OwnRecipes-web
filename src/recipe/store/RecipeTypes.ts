import { Dispatch as ReduxDispatch } from 'redux';
import * as _ from 'lodash-es';

import { NUMBER_UNDEFINED, STRING_UNDEFINED } from '../../common/constants';
import ItemReducerType from '../../common/store/ItemReducerType';
import { GenericItemReducerAction } from '../../common/store/ReduxHelper';
import { PayloadAction } from '../../common/store/redux';
import { slugify } from '../../common/utility';

export interface Quantity {
  numerator?:   number;
  denominator:  number;
  measurement?: string;
}

export interface IngredientInput extends Partial<Quantity> {
  title:        string;

  quantity?:    string;
}

export interface IngredientDto extends Quantity {
  id:           number;
  title:        string;
}
export interface Ingredient extends IngredientDto {
  quantity?:    string;
}
export const toIngredientDto = (obj: Ingredient): IngredientDto => ({
  id:    obj.id,
  title: obj.title,

  numerator:   obj.numerator,
  denominator: obj.denominator,
  measurement: obj.measurement,
});
export const toIngredient = (dto: IngredientDto): Ingredient => ({
  id:    dto.id,
  title: dto.title,

  numerator:   dto.numerator && dto.numerator > 0 ? dto.numerator : undefined,
  denominator: dto.denominator,
  measurement: dto.measurement,
});

export interface IngredientGroupDto {
  title:       string;
  ingredients: Array<IngredientDto>;
}
export interface IngredientGroup {
  slug:        string;
  title:       string;
  ingredients: Array<Ingredient>;
}
export const toIngredientGroupDto = (obj: IngredientGroup): IngredientGroupDto => ({
  title:       obj.title,
  ingredients: obj.ingredients.map(toIngredientDto),
});
export const toIngredientGroup = (dto: IngredientGroupDto): IngredientGroup => ({
  slug:        slugify(dto.title),
  title:       dto.title,
  ingredients: dto.ingredients.map(toIngredient),
});

export interface SubRecipeDto extends Quantity {
  title:       string;
  slug:        string;
  child_recipe_id: number;
  parent_recipe_id?: number;

  measurement?: string;
}

export interface SubRecipe extends Quantity {
  title:       string;
  slug:        string;
  child_recipe_id:   number;
  parent_recipe_id?: number;
  measurement?: string;

  quantity?:   string;
}
export const toSubRecipeDto = (obj: SubRecipe): SubRecipeDto => ({
  title: obj.title,
  slug:  obj.slug,
  child_recipe_id:  obj.child_recipe_id,
  parent_recipe_id: obj.parent_recipe_id,

  measurement: obj.measurement,
  numerator:   obj.numerator,
  denominator: obj.denominator,
});
export const toSubRecipe = (dto: SubRecipeDto): SubRecipe => ({
  title: dto.title,
  slug:  dto.slug,
  child_recipe_id:  dto.child_recipe_id,
  parent_recipe_id: dto.parent_recipe_id,

  measurement: dto.measurement,
  numerator:   dto.numerator,
  denominator: dto.denominator,
});

export interface CourseDto {
  id:    number;
  title: string;
}

export type Course = CourseDto;
export const toCourse = (dto: CourseDto): Course => dto;
export const toCourseDto = (obj: Course): CourseDto => obj;

export interface CuisineDto {
  id:    number;
  title: string;
}

export type Cuisine = CuisineDto;
export const toCuisine = (dto: CuisineDto): Cuisine => dto;
export const toCuisineDto = (obj: Cuisine): CuisineDto => obj;

export interface SeasonDto {
  id:    number;
  title: string;
}

export type Season = SeasonDto;
export const toSeason = (dto: SeasonDto): Season => dto;
export const toSeasonDto = (obj: Season): SeasonDto => obj;

export interface TagDto {
  id:    number;
  title: string;
}

export type Tag = TagDto;

export const toTag = (dto: TagDto): Tag => dto;

export interface RecipeListDto {
  id:    number;
  title: string; // Tasty Chili 24
  slug:  string; // tasty-werwerchili-4

  tags?: Array<TagDto>;

  photo_thumbnail?: string | null;

  info:     string;
  rating:   number;
  rating_count: number;
  pub_date: string; // 2011-05-20
}

export type TagObj = { [key: string]: Tag };
export interface RecipeList {
  id:    number;
  title: string; // Tasty Chili 24
  slug:  string; // tasty-werwerchili-4

  tags?: Array<TagDto>;
  oTags?: TagObj;

  photoThumbnail?: string;

  info:     string;
  rating:   number;
  ratingCount: number;
  pub_date:  string; // 2011-05-20
}

export const toRecipeList = (dto: RecipeListDto): RecipeList => ({
  id:    dto.id,
  title: dto.title,
  slug:  dto.slug,

  tags:  dto.tags?.map(toTag),
  oTags: _.keyBy(dto.tags?.map(toTag) ?? [], 'title'),

  photoThumbnail: dto.photo_thumbnail ?? undefined,

  info:     dto.info,
  rating:   dto.rating,
  ratingCount: dto.rating_count,
  pub_date: dto.pub_date,
});

export interface RecipeDto extends RecipeListDto {
  source:     string;

  cook_time?: number;
  prep_time?: number;
  servings:   number;

  course?:  Course;
  cuisine?: Cuisine;
  season?:  Season;
  tags:     Array<TagDto>;

  photo?: string | null;

  subrecipes: Array<SubRecipeDto>;
  ingredient_groups: Array<IngredientGroupDto>;

  directions: string;

  public: boolean;

  author:       number;
  pub_username: string;
  update_author?:   number;
  update_username?: string;
  update_date:      string;
}

export interface Recipe extends RecipeList {
  source:      string;

  cookTime?: number;
  prepTime?: number;
  servings:  number;

  course?:  Course;
  cuisine?: Cuisine;
  season?:  Season;
  tags:     Array<Tag>;
  oTags:    TagObj;

  photo?: string;

  subrecipes: Array<SubRecipe>;
  ingredientGroups: Array<IngredientGroup>;

  directions: string;

  public: boolean;

  author:       number;
  pub_username: string;
  update_author?:   number;
  update_username?: string;
  update_date:      string;

  customServings: number;
}

export const toRecipe = (dto: RecipeDto): Recipe => ({
  id:    dto.id,
  title: dto.title,
  slug:  dto.slug,

  source:   dto.source,

  cookTime: parseBackendNumber(dto.cook_time),
  prepTime: parseBackendNumber(dto.prep_time),
  servings: dto.servings,

  course:  (dto.course == null || dto.course.title  === '-')  ? undefined : toCourse(dto.course),
  cuisine: (dto.cuisine == null || dto.cuisine.title === '-') ? undefined : toCuisine(dto.cuisine),
  season:  (dto.season == null || dto.season.title  === '-')  ? undefined : toSeason(dto.season),
  tags:    dto.tags.map(toTag),
  oTags:   _.keyBy(dto.tags.map(toTag), 'title'),

  photo: dto.photo ?? undefined,
  photoThumbnail: dto.photo_thumbnail ?? undefined,

  subrecipes: dto.subrecipes.map(toSubRecipe),
  ingredientGroups: dto.ingredient_groups.filter(ig => ig.title !== '-' && ig.ingredients.length > 0).map(toIngredientGroup),

  directions: parseBackendString(dto.directions) ?? '',
  info: dto.info,

  rating: dto.rating,
  ratingCount: dto.rating_count,
  public: dto.public,

  author:       dto.author,
  pub_username: dto.pub_username,
  pub_date:     dto.pub_date,
  update_author:   dto.update_author,
  update_username: dto.update_username,
  update_date:     dto.update_date,

  customServings: dto.servings,
});

export interface RecipeRequest {
  title:      string;

  source:     string;

  cook_time:  number | null;
  prep_time:  number | null;
  servings:   number;

  tags:       Array<TagDto>;
  course:     CourseDto | null;
  cuisine:    CuisineDto | null;
  season:     SeasonDto | null;

  subrecipes: Array<SubRecipeDto>;
  ingredient_groups: Array<IngredientGroupDto>;
  directions: string;
  info:       string;

  photo?:     '';

  public:     boolean;
}

function parseBackendNumber(val: number | undefined): number | undefined {
  if (val == null || val === NUMBER_UNDEFINED) return undefined;
  return val;
}
function parseBackendString(str: string | undefined): string | undefined {
  if (str == null || str === STRING_UNDEFINED) return undefined;
  return str;
}

const toIngredientGroupsDto = (obj: Recipe): Array<IngredientGroupDto> => {
  if (obj.ingredientGroups == null || obj.ingredientGroups.length === 0) {
    if (obj.subrecipes.length > 0) {
      return [{ title: '-', ingredients: [] }];
    } else {
      return [];
    }
  } else {
    return obj.ingredientGroups.filter(ig => ig.ingredients.length > 0).map(toIngredientGroupDto);
  }
};

function ifNull<T>(val: T | undefined, d: T): T {
  if (val == null || (typeof val === 'string' && val === '')) return d;
  return val;
}
export const toRecipeRequest = (obj: Recipe): RecipeRequest => ({
  title:      obj.title,

  source:     ifNull(obj.source, ''),

  cook_time:  obj.cookTime ?? null,
  prep_time:  obj.prepTime ?? null,
  servings:   obj.servings,

  tags:       obj.tags,
  course:     obj.course ? toCourseDto(obj.course) : {} as CourseDto,
  cuisine:    obj.cuisine ? toCuisineDto(obj.cuisine) : {} as CuisineDto,
  season:     obj.season ? toSeasonDto(obj.season) : {} as SeasonDto,

  subrecipes: obj.subrecipes?.map(toSubRecipeDto) ?? [],
  ingredient_groups: toIngredientGroupsDto(obj),
  directions: ifNull(obj.directions, ''),
  info:       ifNull(obj.info, ''),

  photo:      obj.photo === '' ? '' : undefined,

  public:     obj.public,
});

export enum RecipeActionTypes {
  RECIPE_DELETE = 'RECIPE_DELETE',
  RECIPE_INGREDIENT = 'RECIPE_INGREDIENT',
  RECIPE_INGREDIENT_SERVINGS_UPDATE = 'RECIPE_INGREDIENT_SERVINGS_UPDATE',
  RECIPE_LIST_BLANK = 'RECIPE_LIST_BLANK',
  RECIPE_LIST_LOADING = 'RECIPE_LIST_LOADING',
  RECIPE_LIST_COMPLETE = 'RECIPE_LIST_COMPLETE',
  RECIPE_LIST_ERROR = 'RECIPE_LIST_ERROR',
}

export const RECIPE_STORE = 'recipe';

export type IRecipeSlugAction = {
  store: typeof RECIPE_STORE;
  typs:  typeof RecipeActionTypes.RECIPE_DELETE;
} & PayloadAction<{ id: number }>;

export type IRecipeIngredientUpdateServingAction = {
  store: typeof RECIPE_STORE;
  typs:  typeof RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE;
} & PayloadAction<{
  recipeSlug:     string;
  customServings: number;
}>;

export type RecipeState     = ItemReducerType<Recipe>;
export type RecipeAction    = IRecipeSlugAction | IRecipeIngredientUpdateServingAction | GenericItemReducerAction<Recipe>;
export type RecipeDispatch  = ReduxDispatch<RecipeAction>;
