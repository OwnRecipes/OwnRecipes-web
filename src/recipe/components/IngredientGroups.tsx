import '../css/ingredients_panel.css';

import { IngredientGroup } from '../store/RecipeTypes';

import Ingredients from './Ingredients';

export interface IIngredientGroupsProps {
  groups: Array<IngredientGroup> | undefined;
  hasSubrecipes: boolean;
  withHeaderLink?: boolean;
  // checkIngredient: (id: number, checked: boolean) => void;
}

const IngredientGroups: React.FC<IIngredientGroupsProps> = ({
    groups, hasSubrecipes, withHeaderLink /* , checkIngredient */ }: IIngredientGroupsProps) => {
  const showCaptions = hasSubrecipes || (groups != null && groups.length > 1);
  const ingredientGroups = groups?.filter(group => group.title || group.ingredients.length > 0).map(group => (
    <div className='subgroup ingredient-group' key={group.title}>
      <Ingredients
          caption = {showCaptions && group.title ? group.title : undefined}
          withHeaderLink = {withHeaderLink}
          data = {group.ingredients}
          // checkIngredient = {checkIngredient}
          />
    </div>
  ));

  return (
    <>{ingredientGroups}</>
  );
};

export default IngredientGroups;
