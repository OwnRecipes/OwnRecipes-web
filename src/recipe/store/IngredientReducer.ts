import { Dispatch as ReduxDispatch } from 'redux';

import { BasicAction, PayloadAction } from '../../common/store/redux';
import { Ingredient, IngredientGroup, RecipeActionTypes } from './RecipeTypes';

export const RECIPE_INGREDIENTS_STORE = '@@recipeIngredients';

export type QuantityFormatter = (numerator: number | undefined, denominator: number) => string;

export enum RecipeIngredientReducerActionTypes {
  RECIPE_INGREDIENTS_LOAD            = 'RECIPE_INGREDIENTS_LOAD',
}

export type IRecipeIngredientLoadAction = {
  store: typeof RECIPE_INGREDIENTS_STORE;
  typs: typeof RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_LOAD;
  formatQuantity: (numerator: number | undefined, denominator: number) => string;
} & PayloadAction<Array<IngredientGroup>>;

export type IRecipeIngredientServingsUpdateAction = {
  store: typeof RECIPE_INGREDIENTS_STORE;
  typs: typeof RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE;
  formatQuantity: (numerator: number | undefined, denominator: number) => string;
} & BasicAction;

export type RecipeIngredientsState = Array<IngredientGroup>;
export type RecipeIngredientsAction = IRecipeIngredientLoadAction | IRecipeIngredientServingsUpdateAction;
export type RecipeIngredientsDispatch  = ReduxDispatch<RecipeIngredientsAction>;

type RecalcFunction<T> = (ingr: T) => T;

const ingredients = (igs: Array<IngredientGroup>, cb: RecalcFunction<Ingredient>): Array<IngredientGroup> => igs.map(ig => ({
  ...ig,
  ingredients: ig.ingredients.map(ingredient => cb(ingredient)),
}));

// eslint-disable-next-line arrow-body-style
const merge = (state: RecipeIngredientsState, ingredientGroups: Array<IngredientGroup>, formatQuantity: QuantityFormatter) => {
  return ingredients(ingredientGroups, ingr => {
    const custom = formatQuantity(ingr.numerator, ingr.denominator);
    return { ...ingr, quantity: custom };
  });
};

const defaultState: RecipeIngredientsState = [];

const ingredient = (state = defaultState, action: RecipeIngredientsAction): RecipeIngredientsState => {
  if (action.store === RECIPE_INGREDIENTS_STORE) {
    switch (action.typs) {
      case RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_LOAD:
        return merge(state, action.payload, action.formatQuantity);
      case RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE:
        return ingredients(state, i => {
          const custom = action.formatQuantity(i.numerator, i.denominator);
          return { ...i, quantity: custom };
        });
      default:
        break;
    }
  }

  return state;
};

export default ingredient;
