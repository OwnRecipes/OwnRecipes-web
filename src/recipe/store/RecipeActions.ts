import { request } from '../../common/CustomSuperagent';
import { serverURLs } from '../../common/config';
import { ACTION } from '../../common/store/ReduxHelper';
import { toBasicAction } from '../../common/store/redux';
import { handleError } from '../../common/requestUtils';
import { Recipe, RecipeAction, RecipeActionTypes, RecipeDispatch, RecipeDto, RECIPE_STORE, toRecipe } from './RecipeTypes';

export const getRecipeSuccess = (recipe: Recipe): RecipeAction => (
  { ...toBasicAction(RECIPE_STORE, ACTION.GET_SUCCESS), payload: recipe }
);

export const load = (recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({ ...toBasicAction(RECIPE_STORE, ACTION.GET_START) });
  request
    .get(`${serverURLs.recipe}${recipeSlug}/`)
    .then(res => {
      dispatch(getRecipeSuccess(toRecipe(res.body as RecipeDto)));
    })
    .catch(err => dispatch(handleError(err, RECIPE_STORE)));
};

export const deleteRecipe = (id: number, recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({ ...toBasicAction(RECIPE_STORE, ACTION.DELETE_START) });
  request
    .delete(`${serverURLs.recipe}${recipeSlug}/`)
    .then(() => {
      dispatch({ ...toBasicAction(RECIPE_STORE, RecipeActionTypes.RECIPE_DELETE), payload: { id: id } });
    })
    .catch(err => dispatch(handleError(err, RECIPE_STORE)));
};

export const updateServings = (recipeSlug: string, value: number) => (dispatch: RecipeDispatch) => {
  dispatch({
    ...toBasicAction(
      RECIPE_STORE,
      RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE
    ),
    payload: {
      recipeSlug: recipeSlug,
      customServings: value,
    },
  });
};

export const preload = (recipe: Partial<Recipe>) => (dispatch: RecipeDispatch) => {
  dispatch({ ...toBasicAction(RECIPE_STORE, ACTION.PRELOAD), payload: recipe });
};

export const reset = () => (dispatch: RecipeDispatch) => {
  dispatch({ ...toBasicAction(RECIPE_STORE, ACTION.RESET) });
};
