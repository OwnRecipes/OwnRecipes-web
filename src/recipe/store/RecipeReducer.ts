import * as _ from 'lodash-es';

import ingredient, { RecipeIngredientReducerActionTypes, RECIPE_INGREDIENTS_STORE } from './IngredientReducer';
import subrecipes, { RecipeSubrecipesReducerActionTypes, RECIPE_SUBRECIPES_STORE } from './SubRecipeReducer';
import fq from '../utilts/formatQuantity';
import ReduxHelper, { ACTION, GenericItemReducerAction } from '../../common/store/ReduxHelper';
import { Recipe, RecipeAction, RecipeActionTypes, RecipeState, RECIPE_STORE } from './RecipeTypes';
import { toBasicAction } from '../../common/store/redux';

const defaultState: RecipeState = ReduxHelper.getItemReducerDefaultState<Recipe>(RECIPE_STORE);

const recipe = (state = defaultState, action: RecipeAction): RecipeState => {
  if (action.store === RECIPE_STORE) {
      switch (action.typs) {
      case ACTION.GET_SUCCESS:
        {
          const actionRecipe = action.payload;
          if (!actionRecipe) return ReduxHelper.setItem<Recipe, RecipeState>(state, undefined);

          const isNew = state.item == null || state.item.id !== actionRecipe.id || state.item.ingredientGroups == null;
          const updServings = isNew ? actionRecipe.servings : (state.item?.customServings ?? 1);
          const subRecipes = subrecipes(
            [],
            {
              ...toBasicAction(
                RECIPE_SUBRECIPES_STORE,
                RecipeSubrecipesReducerActionTypes.RECIPE_SUBRECIPES_LOAD
              ),
              payload: actionRecipe.subrecipes,
              // OPT this is magic
              // just pass updServings and actionRecipe.servings via the payload
              formatQuantity: fq.bind(this, updServings, actionRecipe.servings),
            }
          );
          const ingredients = ingredient(
            [],
            {
              ...toBasicAction(
                RECIPE_INGREDIENTS_STORE,
                RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_LOAD
              ),
              payload: actionRecipe.ingredientGroups,
              formatQuantity: fq.bind(this, updServings, actionRecipe.servings),
            }
          );

          const updItem: Recipe = {
            ...actionRecipe,
            subrecipes:       subRecipes,
            ingredientGroups: ingredients,
            customServings:   updServings,
          };
          return ReduxHelper.setItem(state, updItem);
        }
      case RecipeActionTypes.RECIPE_DELETE:
        return defaultState;
      case RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE:
        {
          if (state.item == null) return state;

          const { customServings } = action.payload;
          let updItem: Recipe = _.clone(state.item);

          const subRecipes = subrecipes(
            updItem.subrecipes,
            {
              ...toBasicAction(
                RECIPE_SUBRECIPES_STORE,
                RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE
              ),
              formatQuantity: fq.bind(this, updItem.servings, customServings),
            }
          );
          const ingredients = ingredient(
            updItem.ingredientGroups,
            {
              ...toBasicAction(
                RECIPE_INGREDIENTS_STORE,
                RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE
              ),
              formatQuantity: fq.bind(this, updItem.servings, customServings),
            }
          );
          updItem = {
            ...updItem,
            subrecipes: subRecipes,
            ingredientGroups: ingredients,
            customServings: customServings,
          };

          return ReduxHelper.setItem(state, updItem);
        }
      default: break;
    }
  }

  return ReduxHelper.caseItemDefaultReducer(state, action as GenericItemReducerAction<Recipe>, defaultState);
};

export default recipe;
