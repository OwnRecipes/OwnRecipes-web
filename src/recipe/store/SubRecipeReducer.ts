import { Dispatch as ReduxDispatch } from 'redux';
import { BasicAction, PayloadAction } from '../../common/store/redux';
import { QuantityFormatter } from './IngredientReducer';
import { RecipeActionTypes, SubRecipe } from './RecipeTypes';

export const RECIPE_SUBRECIPES_STORE = '@@recipeSubrecipes';

export enum RecipeSubrecipesReducerActionTypes {
  RECIPE_SUBRECIPES_LOAD = 'RECIPE_SUBRECIPES_LOAD',
}

export type IRecipeSubrecipesLoadAction = {
  store: typeof RECIPE_SUBRECIPES_STORE;
  typs: typeof RecipeSubrecipesReducerActionTypes.RECIPE_SUBRECIPES_LOAD;
  formatQuantity: QuantityFormatter;
} & PayloadAction<Array<SubRecipe>>;

export type IRecipeSubrecipesIngredientsUpdateAction = {
  store: typeof RECIPE_SUBRECIPES_STORE;
  typs: typeof RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE;
  formatQuantity: QuantityFormatter;
} & BasicAction;

export type RecipeSubrecipesState    = Array<SubRecipe>;
export type RecipeSubrecipesAction   = IRecipeSubrecipesLoadAction | IRecipeSubrecipesIngredientsUpdateAction;
export type RecipeSubrecipesDispatch = ReduxDispatch<RecipeSubrecipesAction>;

// eslint-disable-next-line arrow-body-style
const merge = (state: RecipeSubrecipesState, subrecipes: Array<SubRecipe>, formatQuantity: QuantityFormatter) => {
  return subrecipes.map(subrecipe => {
    const custom = formatQuantity(subrecipe.numerator, subrecipe.denominator);
    return { ...subrecipe, quantity: custom };
  });
};

const defaultState: RecipeSubrecipesState = [];

const subRecipes = (state = defaultState, action: RecipeSubrecipesAction) => {
  if (action.store === RECIPE_SUBRECIPES_STORE) {
    switch (action.typs) {
      case RecipeSubrecipesReducerActionTypes.RECIPE_SUBRECIPES_LOAD:
        return merge(state, action.payload, action.formatQuantity);
      case RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE:
        return state.map(i => {
          const custom = action.formatQuantity(i.numerator, i.denominator);
          return { ...i, quantity: custom };
        });
      default:
        break;
    }
  }

  return state;
};

export default subRecipes;
