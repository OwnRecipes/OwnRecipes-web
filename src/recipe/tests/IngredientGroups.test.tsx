import IngredientGroups from '../components/IngredientGroups';
import create from '../../test/create';

import data from './data';
import { IngredientGroup } from '../store/RecipeTypes';

test('Ingredient Group component test', () => {
  // const checkIngredient = jest.fn();
  const component = create(
    <IngredientGroups groups={data.ingredient_groups as unknown as Array<IngredientGroup>} hasSubrecipes={false} withHeaderLink /* checkIngredient={checkIngredient} */ />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
