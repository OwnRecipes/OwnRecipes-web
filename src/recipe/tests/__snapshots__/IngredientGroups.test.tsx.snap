// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingredient Group component test 1`] = `
Array [
  <div
    className="subgroup ingredient-group"
  >
    <table
      className="table ingredients-table table table-sm table-striped"
    >
      <caption
        className="subheading h3"
        id="ingredients-chili"
      >
        Chili:
        <a
          className="headerlink"
          href="/#ingredients-chili"
          onClick={[Function]}
          title="Permalink to this headline"
        >
          ¶
        </a>
      </caption>
      <thead
        className="hideme"
      >
        <tr>
          <th>
            <span>
              Quantity
            </span>
          </th>
          <th>
            <span>
              Ingredient
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="ingredient"
        >
          <td
            className="quantity"
          >
            <span>
              1
               
              dash
            </span>
          </td>
          <td
            className="ingredient"
          >
            <span>
              black pepper
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="subgroup ingredient-group"
  >
    <table
      className="table ingredients-table table table-sm table-striped"
    >
      <caption
        className="subheading h3"
        id="ingredients-spices"
      >
        Spices:
        <a
          className="headerlink"
          href="/#ingredients-spices"
          onClick={[Function]}
          title="Permalink to this headline"
        >
          ¶
        </a>
      </caption>
      <thead
        className="hideme"
      >
        <tr>
          <th>
            <span>
              Quantity
            </span>
          </th>
          <th>
            <span>
              Ingredient
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="ingredient"
        >
          <td
            className="quantity"
          >
            <span>
              1
               
              dash
            </span>
          </td>
          <td
            className="ingredient"
          >
            <span>
              black pepper
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="subgroup ingredient-group"
  >
    <table
      className="table ingredients-table table table-sm table-striped"
    >
      <thead
        className="hideme"
      >
        <tr>
          <th>
            <span>
              Quantity
            </span>
          </th>
          <th>
            <span>
              Ingredient
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="ingredient"
        >
          <td
            className="quantity"
          >
            <span>
              1
               
              dash
            </span>
          </td>
          <td
            className="ingredient"
          >
            <span>
              black pepper
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
